electrum/payment_identifier.py

```diff
@@ -0,0 +1,768 @@
+        if is_multiline and errors:
+            self.error = errors.strip() if errors else None
+        self.logger.debug(f'multiline: {outputs!r}, {self.error}')
+        return outputs
+
+    def parse_address_and_amount(self, line) -> 'PartialTxOutput':
+        try:
+            x, y = line.split(',')
+        except ValueError:
+            raise Exception("expected two comma-separated values: (address, amount)") from None
+        scriptpubkey = self.parse_output(x)
+        if not scriptpubkey:
+            raise Exception('Invalid address')
+        amount = self.parse_amount(y)
+        return PartialTxOutput(scriptpubkey=scriptpubkey, value=amount)
+
+    def parse_output(self, x) -> bytes:
+        try:
+            address = self.parse_address(x)
+            return bytes.fromhex(bitcoin.address_to_script(address))
+        except Exception as e:
+            pass
+        try:
+            script = self.parse_script(x)
+            return bytes.fromhex(script)
+        except Exception as e:
+            pass
+
+        # raise Exception("Invalid address or script.")
+
+    def parse_script(self, x):
+        script = ''
+        for word in x.split():
+            if word[0:3] == 'OP_':
+                opcode_int = opcodes[word]
+                script += construct_script([opcode_int])
+            else:
+                bytes.fromhex(word)  # to test it is hex data
+                script += construct_script([word])
+        return script
+
+    def parse_amount(self, x):
+        x = x.strip()
+        if not x:
+            raise Exception("Amount is empty")
+        if parse_max_spend(x):
+            return x
+        p = pow(10, self.config.get_decimal_point())
+        try:
+            return int(p * Decimal(x))
+        except InvalidOperation:
+            raise Exception("Invalid amount")
+
+    def parse_address(self, line):
+        r = line.strip()
+        m = re.match('^' + RE_ALIAS + '$', r)
+        address = str(m.group(2) if m else r)
+        assert bitcoin.is_address(address)
+        return address
+
+    def get_fields_for_GUI(self) -> FieldsForGUI:
+        recipient = None
+        amount = None
+        description = None
+        validated = None
+        comment = None
+        amount_range = None
+
+        if (self.emaillike or self.domainlike) and self.openalias_data:
+            key = self.emaillike if self.emaillike else self.domainlike
+            address = self.openalias_data.get('address')
+            name = self.openalias_data.get('name')
+            description = name
+            recipient = key + ' <' + address + '>'
+            validated = self.openalias_data.get('validated')
+            if not validated:
+                self.warning = _('WARNING: the alias "{}" could not be validated via an additional '
+                                 'security check, DNSSEC, and thus may not be correct.').format(key)
+
+        elif self.bolt11 and self.wallet.has_lightning():
```

Comment:
I think the output of this function should not depend on `wallet.has_lightning()`. If a wallet does not support LN, it should handle that itself.

I don't like that `PaymentIdentifier` now has a reference to wallet, just checking where it is used and what for.

Reply:
> I think the output of this function should not depend on `wallet.has_lightning()`. If a wallet does not support LN, it should handle that itself.

Agreed. Also, lnurl type payment identifiers  do not check for this either.

> I don't like that `PaymentIdentifier` now has a reference to wallet, just checking where it is used and what for.

Basically only `wallet.config` and `wallet.contacts` are used.

Reply:
no, there is also get_invoice

Reply:
I added a FIXME in the code regarding that.